File: apps/api/src/dto.rs
use serde::Serialize;
use serde_json::Value;

#[derive(Debug, Clone, Serialize)]
pub struct TraceCard {
    pub id: String,
    pub service: String,
    pub root_service: String,
    pub name: String,
    pub duration_ms: i64,
    pub span_count: u32,
    pub status: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub http_status_code: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub http_url: Option<String>,
    pub start_time: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct TraceFacetsDto {
    pub services: Vec<String>,
    pub statuses: Vec<String>,
    pub methods: Vec<String>,
    pub http_status_codes: Vec<i32>,
    pub routes: Vec<RouteFacetDto>,
}

#[derive(Debug, Clone, Serialize)]
pub struct RouteFacetDto {
    pub value: String,
    pub count: u64,
}

#[derive(Debug, Clone, Serialize)]
pub struct SpanDto {
    pub id: String,
    pub parent_id: Option<String>,
    pub name: String,
    pub service: String,
    pub kind: i32,
    pub status: String,
    pub status_message: Option<String>,
    pub start_offset_ms: f64,
    pub duration_ms: f64,
    pub attributes: Value,
    pub events: Value,
    pub links: Value,
    pub resource_attributes: Value,
    pub scope_name: Option<String>,
    pub scope_version: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
pub struct TraceDetailDto {
    pub trace: TraceCard,
    pub spans: Vec<SpanDto>,
}

#[derive(Debug, Clone, Serialize)]
pub struct ServiceCard {
    pub name: String,
    pub trace_count: u64,
}

#[derive(Debug, Clone, Serialize)]
pub struct LogDto {
    pub id: String,
    pub time: String,
    pub severity_number: Option<i32>,
    pub severity_text: Option<String>,
    pub body: Value,
    pub service_name: String,
    pub attributes: Value,
    pub scope_name: Option<String>,
    pub scope_version: Option<String>,
    pub trace_id: Option<String>,
    pub span_id: Option<String>,
}

File: apps/api/src/features/traces/types/dto.ts
import type { Json } from "@shared/helpers"

export type TraceCardDto = {
  id: string
  service: string
  root_service: string
  name: string
  duration_ms: number
  span_count: number
  status: string
  http_status_code?: number
  http_method?: string
  http_url?: string
  http_route?: string
  start_time: string
  breakdown: BreakdownDto[] | null
}

export type BreakdownDto = {
  name: string
  duration_ms: number
  share: number
}

export type FacetValueDto = {
  value: string
  count: number
}

export type TraceFacetsDto = {
  services: FacetValueDto[]
  statuses: FacetValueDto[]
  methods: FacetValueDto[]
  http_status_codes: FacetValueDto[]
  routes: FacetValueDto[]
  durations: FacetValueDto[]
}

export type SpanDto = {
  id: string
  parent_id: string | null
  name: string
  service: string
  kind: number
  status: string
  status_message: string | null
  start_offset_ms: number
  duration_ms: number
  attributes: Json
  events: Json
  links: Json
  resource_attributes: Json
  scope_name: string | null
  scope_version: string | null
  type?: string
  payload_path?: string
  provider: string
}

export type TraceDetailDto = {
  trace: TraceCardDto
  spans: SpanDto[]
}

export type SpanOverviewDto = {
  id: string
  parent_id: string | null
  name: string
  service: string
  type?: string
  status: string
  start_offset_ms: number
  duration_ms: number
  payload_path?: string
}

export type TraceOverviewDto = {
  trace: TraceCardDto
  spans: SpanOverviewDto[]
  counts: {
    spans: number
    errors: number
    by_type: Record<string, number>
  }
  hints: string[]
}

export type TypedSpanDto = {
  span_id: string
  trace_id: string
  name: string
  type: string
  service: string
  status: string
  duration_ms: number
  start_offset_ms: number
  payload_path?: string
  payload: Record<string, unknown>
}

export type SqlQueryDto = {
  span_id: string
  name: string
  statement: string
  duration_ms: number
  start_offset_ms: number
  started_at: string | null
  db_system: string | null
  host: string | null
  status: string
  share: number
}
